azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/compute/extensions/AzureComputeImageExtension.java

```diff
@@ -0,0 +1,135 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jclouds.azurecompute.arm.compute.extensions;
+
+import com.google.common.base.Predicate;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import com.google.gson.internal.LinkedTreeMap;
+import com.google.inject.Inject;
+import com.google.inject.name.Named;
+import org.jclouds.Constants;
+import org.jclouds.azurecompute.arm.AzureComputeApi;
+import org.jclouds.azurecompute.arm.compute.config.AzureComputeServiceContextModule;
+import org.jclouds.azurecompute.arm.compute.functions.VMImageToImage;
+import org.jclouds.azurecompute.arm.domain.ResourceDefinition;
+import org.jclouds.azurecompute.arm.domain.VMImage;
+import org.jclouds.azurecompute.arm.domain.VirtualMachine;
+import org.jclouds.compute.domain.CloneImageTemplate;
+import org.jclouds.compute.domain.Image;
+import org.jclouds.compute.domain.ImageTemplate;
+import org.jclouds.compute.domain.ImageTemplateBuilder;
+import org.jclouds.compute.extensions.ImageExtension;
+import org.jclouds.util.Predicates2;
+
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Callable;
+
+
+public class AzureComputeImageExtension implements ImageExtension {
+   private final AzureComputeApi api;
+   private final ListeningExecutorService userExecutor;
+   private final String group;
+   private final VMImageToImage imageReferenceToImage;
+   public static final String CONTAINER_NAME = "vhdsnew";
+   public static final String CUSTOM_IMAGE_PREFIX = "#";
+
+   @Inject
+   AzureComputeImageExtension(AzureComputeApi api, final AzureComputeServiceContextModule.AzureComputeConstants azureComputeConstants,
+                              @Named(Constants.PROPERTY_USER_THREADS) ListeningExecutorService userExecutor,
+                              VMImageToImage imageReferenceToImage) {
+      this.userExecutor = userExecutor;
+      this.group = azureComputeConstants.azureResourceGroup();
+      this.imageReferenceToImage = imageReferenceToImage;
+      this.api = api;
+   }
+
+   @Override
+   public ImageTemplate buildImageTemplateFromNode(String name, String id) {
+      name = name.toLowerCase();
+      return new ImageTemplateBuilder.CloneImageTemplateBuilder().nodeId(id).name(name).build();
+   }
+
+   @Override
+   public ListenableFuture<Image> createImage(ImageTemplate template) {
+      final CloneImageTemplate cloneTemplate = (CloneImageTemplate) template;
+      String id = cloneTemplate.getSourceNodeId();
+      final String storageAccountName = id.replaceAll("[^A-Za-z0-9 ]", "") + "stor";
+
+      boolean generalized = false;
+      while (!generalized) {
```

Comment:
Hmm... if I correctly remember, there was some timing issue and first call was not always working, but second call has been working if first did not :) I guess we can leave it like this or then we should just remove whole while loop and trust that it works without that (as I cannot reproduce that issue now).